### PR TITLE
[feature] Add pinot-clp-log plugin to encode user-specified fields with CLP during ingestion

### PR DESCRIPTION
**Pull request by kirkrodrigues**
At a high-level, the plugin takes two inputs: a JSON record and a list of fields (unstructured log messages) to encode with [CLP](https://github.com/y-scope/clp). The plugin will extract and encode the user-specified fields into CLP's three-column format and store the output in a Pinot `GenericRow` object.

This is part of the change requested in #9819 and described in this [design doc](https://docs.google.com/document/d/1nHZb37re4mUwEA258x3a2pgX13EWLWMJ0uLEDk1dUyU).

# Release notes
* New plugin added: `pinot-clp-log` to encode user-specified fields with CLP during ingestion.
* Users can use the plugin by specifying these configuration options in their `tableIndexConfig.streamConfigs`:

  ```json
  "stream.kafka.decoder.class.name": "org.apache.pinot.plugin.inputformat.clplog.CLPLogMessageDecoder",
  "stream.kafka.decoder.prop.fieldsForClpEncoding": "<field-names>"
  ```
   
   where `<field-names>` is a comma-separated list of fields you wish to encode with CLP.

# Testing performed
* Validated fields are encoded correctly using the added unit test.
* Created a table with the following settings:

```json
"tableIndexConfig": {
    ...,
    "streamConfigs": {
        ...,
        "stream.kafka.decoder.class.name": "org.apache.pinot.plugin.inputformat.clplog.CLPLogMessageDecoder",
        "stream.kafka.decoder.prop.fieldsForClpEncoding": "message",
    }
}
```

* Ingested logs through Kafka and ensured log events containing the message field were transformed such that the `message` field was replaced with CLP's three fields: `message_logtype`, `message_dictionaryVars`, and `message_encodedVars`.

### CODE REVIEW

**Comment by Jackie-Jiang** [comment]
This is a great feature! Can you please give public access to the design doc?

**Comment by kirkrodrigues** [comment]
> This is a great feature! Can you please give public access to the design doc?

Apologies for the delay, the previous doc couldn't be shared due to some security settings. I've updated the link to a publicly accessible doc.

File: pinot-plugins/pinot-input-format/pinot-clp-log/src/test/java/org/apache/pinot/plugin/inputformat/clplog/CLPLogRecordExtractorTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotEquals;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.fail;
+
+
+public class CLPLogRecordExtractorTest {
```
Comment (chenboat) [inline pinot-plugins/pinot-input-format/pinot-clp-log/src/test/java/org/apache/pinot/plugin/inputformat/clplog/CLPLogRecordExtractorTest.java:39]: why is this an extractor test when it tests and decoder actually? Can we test the extractor directly?

File: pinot-plugins/pinot-input-format/pinot-clp-log/src/test/java/org/apache/pinot/plugin/inputformat/clplog/CLPLogRecordExtractorTest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    fieldsToRead.add("message1_logtype");
+    fieldsToRead.add("message1_encodedVars");
+    fieldsToRead.add("message1_dictionaryVars");
+    fieldsToRead.add("message2_logtype");
+    fieldsToRead.add("message2_encodedVars");
+    fieldsToRead.add("message2_dictionaryVars");
+    // Add an unencoded field
+    fieldsToRead.add("timestamp");
+    try {
+      messageDecoder.init(props, fieldsToRead, null);
```
Comment (chenboat) [inline pinot-plugins/pinot-input-format/pinot-clp-log/src/test/java/org/apache/pinot/plugin/inputformat/clplog/CLPLogRecordExtractorTest.java:57]: We also need to test different scenarios based on props's value.

File: pinot-distribution/pinot-assembly.xml [hunk 3]
```diff
@@ -146,6 +146,12 @@
       </source>
       <destName>plugins/pinot-input-format/pinot-avro/pinot-avro-${project.version}-shaded.jar</destName>
     </file>
+    <file>
+      <source>
+        ${pinot.root}/pinot-plugins/pinot-input-format/pinot-clp-log/target/pinot-clp-log-${project.version}-shaded.jar
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply pinot-distribution/pinot-assembly.xml:151]: No, I'm not seeing a version conflict, but aren't all plugins shaded? If you're asking why I had shading in the plugin's pom.xml, that was an error (thanks for catching it).

**Review by chenboat — APPROVED** [review]

**Comment by Jackie-Jiang** [comment]
Thanks for adding this powerful feature. Can you help add a documentation page in the [pinot doc](https://github.com/pinot-contrib/pinot-docs) describing when and how to use the feature?

**Comment by kirkrodrigues** [comment]
Yep, will do.

**Pull request merged**